FAIRmat 2024: additional base classes in NXinstrument

lukaspie opened the pull request:
This PR adds four unnamed base classes to `NXinstrument`.
- `NXactuator`: see description/discussion in #1414
- `NXfabrication`: see description/discussion in #1414  
- `NXhistory`: this is a new base class that is meant to describe the history of any physical object (in this case, the instrument history).  `NXhistory` is meant to contain many instances of the new base class `NXactivity`, which describes any physical activity that was performed on the physical entity prior or during the experiment. Note this is intentionally different than `NXprocess`, which describes processing of the _data_.
- `NXresolution`: This is meant to describe the resolution of a measureable physical quantity. Here, we are describing many different ways of a) how such a resolution can be described, and b) how such a resolution can be obtained. For b), we are also moving the base class `NXcalibration` (a subclass of `NXprocess`) from contributed to base_classes and slightly expand its former definition .

In addition, `NXnote` is expanded to contain more information about a given resource (like the URI of the resource,  the checksum of a file, etc.).

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+        <doc>
+             Any activity that was performed on the physical entity prior or during the experiment. In
+             the future, if there is base class inheritance, this can describe any activity,
+             including processes and measurements.
```
Remove the comment on the future

lukaspie replied:
done

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+subentry(NXsuxbentry):
+doc: |
+Any activity that was performed on the physical entity prior or during the experiment.
+definition: ["NXactivity"]-->
```
Remove comment

lukaspie replied:
done

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+        <doc>
+             Any chemical process that was performed on the physical entity prior or during the
+             experiment.
+        </doc>
```
Remove physical and chemical process as that will be covered by the NXactivity

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+             persistent) identifier of e.g. another file which gives
+             as many as possible details of the history event.
+        </doc>
+    </group>
```
Remove NXidentifier

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+             as many as possible details of the history event.
+        </doc>
+    </group>
+    <!--There should be more activities here, like measurement.-->
```
Remove comment

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+             reference to the location or a unique identifier or other metadata file. In the
+             case these are not available, free-text description.
+             This should only be used in case that there is no rigorous description
+             using the base classes above. This field can also be used to pull in any activities
```
```suggestion
             using the base classes above. This group can also be used to pull in any activities
```

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+        </doc>
+    </group>
+    <!--There should be more activities here, like measurement.-->
+    <group name="notes" type="NXnote">
```
Remove name="notes"

phyy-nx commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+             case these are not available, free-text description.
+             This should only be used in case that there is no rigorous description
+             using the base classes above. This field can also be used to pull in any activities
+             that are not well described by an existing base class definition.
```
Add comment that any number of NXnotes are allowed for extra details.

lukaspie replied:
done

phyy-nx commented on base_classes/NXactivity.nxdl.xml:
```diff
@@ -0,0 +1,56 @@
+         
+         This class is planned be used in the future as the super class for all other activities if inheritance 
+         in base classes is supported in NeXus.
+    </doc>
```
Remove reference to the future.  "This class is a super class..."

phyy-nx commented on base_classes/NXactivity.nxdl.xml:
```diff
@@ -0,0 +1,56 @@
+    </doc>
+    <field name="start_time" type="NX_DATE_TIME">
+        <doc>
+             ISO 8601 formatted time code (with local time zone offset to UTC information
```
Remove ISO reference.  "Start time of activity"

phyy-nx commented on base_classes/NXactivity.nxdl.xml:
```diff
@@ -0,0 +1,56 @@
+             Short description of the activity.
+        </doc>
+    </field>
+    <group name="notes" type="NXnote">
```
Remove name="notes"

phyy-nx commented on base_classes/NXactivity.nxdl.xml:
```diff
@@ -0,0 +1,56 @@
+             ISO 8601 formatted time code (with local time zone offset to UTC information
+             included) when this activity ended.
+        </doc>
+    </field>
```
Add `end_time_estimated` or add an estimated attribute on `end_time`.

phyy-nx commented on base_classes/NXactivity.nxdl.xml:
```diff
@@ -0,0 +1,56 @@
+             This can be any data or other descriptor acquired during the activity
+             (NXnote allows to add pictures, audio, movies). Alternatively, a
+             reference to the location or a unique identifier or other metadata file. In the
+             case these are not available, free-text description.
```
Add comment that any number of NXnotes are allowed for extra details.

phyy-nx commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+    </field>
+    <field name="relative_resolution" type="NX_FLOAT" units="NX_ANY">
+        <doc>
+             Ratio of the resolution at a specified measurand value to that measurand value,
```
```suggestion
             Ratio of the resolution at a specified measurand value to that measurand value
```

lukaspie commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+    </group>
+    <field name="formula_SYMBOL" type="NX_CHAR">
+        <doc>
+             A symbol linking to another path in this appdef to be referred to from the
```
This should be "A symbol linking to another path in the NeXus tree to be referred to from the..."

phyy-nx replied:
spell out appdef

woutdenolf commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+    <field name="physical_quantity">
+        <doc>
+             The physical quantity of the resolution, e.g.,
+             energy, momentum, time, etc.
```
Can we include spatial resolution as an example (e.g. imaging techniques)?

phyy-nx commented on base_classes/NXinstrument.nxdl.xml:
```diff
@@ -55,16 +56,20 @@
     <group type="NXmonochromator" />
     <group type="NXpolarizer" />
     <group type="NXpositioner" />
+    <group type="NXresolution"/>
```
Remove NXresolution from NXinstrument

woutdenolf commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+    </field>
+    <field name="relative_resolution" type="NX_FLOAT" units="NX_ANY">
+        <doc>
+             Ratio of the resolution at a specified measurand value to that measurand value,
```
measurand -> measured?

phyy-nx replied:
Telco: measurand is fine

phyy-nx commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+    </field>
+    <group name="response_function" type="NXdata">
+        <doc>
+             The response of the instrument or part to a infinitesimally sharp input signal
```
```suggestion
             The response of the instrument or part of the instrument to a infinitesimally sharp input signal
```

phyy-nx commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+                 in the input axis or grid.
+                 This field should have the same dimensions as `input`.
+            </doc>
+        </field>
```
Remove input and magnitude, and expand doc string to describe how this should be used.

phyy-nx commented on base_classes/NXresolution.nxdl.xml:
```diff
@@ -0,0 +1,125 @@
+             entered by the `formula_...` fields.
+             The output unit should match the provided unit of this field.
+        </doc>
+    </field>
```
Remove formula_SYMBOL and change resolution_formula to resolution_description, update doc to indicate an english description of the math used.

phyy-nx replied:
Actually use NXparameters for formula_SYMBOL

lukaspie commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+             calibration method but no actual calibration data.
+        </doc>
+    </group>
+    <group name="calibration_reference" type="NXidentifier">
```
to be replaced by identifier attribute

phyy-nx replied:
Waiting on resolution of #1486

lukaspie commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+    </field>
+    <field name="mapping_MAPPING" type="NX_FLOAT">
+        <doc>
+             Mapping data for calibration.
```
example: transmission function in XPS

lukaspie commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+            <dim index="1" value="ncoeff"/>
+        </dimensions>
+    </field>
+    <field name="fit_function" type="NX_CHAR">
```
Following discussion in NXresolution, a similar compromise would be ok for me

lukaspie replied:
Probably also remove attribute `symbol` in `input_data`

phyy-nx replied:
Use formula description instead of formula

lukaspie replied:
Using `NXparameters` and formula description from now on

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+        <dimensions rank="1">
+            <dim index="1" value="ncoeff"/>
+        </dimensions>
+    </field>
```
Use NXparameters instead of coeffiencts

lukaspie replied:
done

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+             name `input_my_field` it should be referred here by `my_field` or `my_field0` if
+             you want to read the zeroth element of the array.
+             
+             The formula should be numpy compliant.
```
Remove numpy

lukaspie replied:
done

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+             The formula should be numpy compliant.
+        </doc>
+    </field>
+    <field name="scaling" type="NX_FLOAT" units="NX_ANY">
```
Use scaling_factor and grab the docstrings for scaling_factor and offset from NXdata

phyy-nx replied:
Would mean correcting the formula to the one used in NXdata

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+        <dimensions rank="1">
+            <dim index="1" value="ncal"/>
+        </dimensions>
+        <attribute name="output_path">
```
Drop output_path

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+        </symbol>
+    </symbols>
+    <doc>
+         Subclass of NXprocess to describe post-processing calibrations.
```
Make it a subclass

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+             Has the calibration been applied?
+        </doc>
+    </field>
+    <field name="calibration_software">
```
Remove due to no being a subclass

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+        <doc>
+             Name of the software used for this calibration.
+        </doc>
+        <attribute name="version">
```
Remove due to now being a subclass

phyy-nx commented on base_classes/NXserialized.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+         from which pieces of information have been extracted for consumption in e.g. a
+         research data management system (RDMS). This may be for reasons of enabling
+         services such as providing access to normalized information for which reading
+         again from the resource may not be desired, possibe, or feasible.
```
```suggestion
         again from the resource may not be desired, possible, or feasible.
```

phyy-nx commented on base_classes/NXserialized.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+    </doc>
+    <field name="type" type="NX_CHAR">
+        <doc>
+             Answers into what resource the information was serialized.
```
```suggestion
             Describes what was serialized
```

tacaswell commented on base_classes/NXserialized.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+            <item value="database"/>
+        </enumeration>
+    </field>
+    <field name="path" type="NX_CHAR">
```
```suggestion
    <field name="uri" type="NX_CHAR">
```

phyy-nx commented on base_classes/NXserialized.nxdl.xml:
```diff
@@ -0,0 +1,74 @@
+             Extracted file containing the serialized information.
+        </doc>
+    </group>
+</definition>
```
Add the desired parameters to NXnote instead

lukaspie replied:
done, removed NXserialized

phyy-nx commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+             The axis values may be copied or linked in the appropriate NXcalibration fields for reference.
+        </doc>
+    </group>
+    <group name="calibration_object" type="NXserialized">
```
Use NXnote instead

lukaspie commented:
Dev notes to keep track:

- NXhistory: implemented suggested edits
- NXactivity: implemented suggested edits
- NXresolution: implemented suggested edits
- NXcalibration: ok following edits
- NXserialization: removed -> moved needed parameters to NXnote
- NXinstrument: implemented suggested edits

depends on / blocked by:

- `identifierNAME` in `NXobject` -> #1486
- implementation of `nameType` in groups -> #1485

ToDo:
- NXcalibration
  - Use `NXparameters` in `formula_SYMBOL`
  - agree on `formula_description`

lukaspie commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,220 @@
+                 Should be a valid NeXus path name, e.g., /entry/instrument/detector/raw.
+            </doc>
+        </attribute>
+    </field>
```
done

phyy-nx commented:
@sanbrock @lukaspie can you write a description at the top of this PR?  Thanks!

lukaspie commented:
Update: I fixed the build (previously failing on links in `NXcalibration`). So once #1486 is ready, there will be very small changes to this PR here and then it should be reviewable immediately.

lukaspie commented:
@phyy-nx this PR should be ready for final review (already voted on).

phyy-nx commented:
For reference, the approving vote is this comment:
https://github.com/nexusformat/definitions/issues/1452#issuecomment-2382668722

Blocked by #1414

lukaspie commented:
> For reference, the approving vote is this comment: [#1452 (comment)](https://github.com/nexusformat/definitions/issues/1452#issuecomment-2382668722)
> 
> Blocked by #1414

Sorry, I missed this was still blocked by #1414. Since that one has been merged, I rebased here, so should be ready to go now.

phyy-nx commented on base_classes/NXnote.nxdl.xml:
```diff
@@ -44,6 +44,27 @@
 	<field name="file_name">
 		<doc>Name of original file name if note was read from an external source</doc>
 	</field>
+	<field name="uri" type="NX_CHAR">
```
One last question, do you still want uri here given that #1486 is now available?

lukaspie replied:
I agree, let's just use `identifierNAME` here as well.

I changed it (but kept the docstring here) and also changed the relevant part of the checksum docs.

lukaspie replied:
Also, I rebased on top of the main branch (which no changes to the files touched here), in case you were wondering where all the commits below are coming from.

lukaspie commented:
@phyy-nx I also made one more change in that I changed the links in the `NXcalibration` docs to point correctly to the concepts in `NXdata`:
![image](https://github.com/user-attachments/assets/09a87385-ec27-42f8-93dc-63433d154a7e)

mkuehbach commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+             energy, momentum, time, etc.
+        </doc>
+    </field>
+    <attribute name="identifier_calibration_method">
```
This cannot take advantage of identifierNAME and specifically the hint on what/who is the data resolver, the docstring just states DOI, or, or, or, one idea on this could be (but agree if at all a new PR):
allow that in xsd a field is allowed to have an attribute for the resolver, so if an NX_CHAR field is found with that attribute it hints on how to resolve what more specifically, this avoids the needs for creating a group
calibration_method(NXobject)
   identifierNAME

lukaspie replied:
you are right, I changed it

mkuehbach commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+    </attribute>
+    <group name="calibration_object" type="NXnote">
+        <doc>
+             A file serialisation of a calibration which may not be publicly available (externally from the nexus file).
```
NeXus file

lukaspie replied:
changed, thanks

phyy-nx commented:
Ready for final review and merge.

lukaspie commented:
Hey @PeterC-DLS, any update on the final review for this one? Thanks a lot!

PeterC-DLS commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+    </field>
+    <field name="original_axis" type="NX_FLOAT" units="NX_ANY">
+        <doc>
+             Vector containing the data coordinates in the original uncalibrated axis
```
```suggestion
             Array containing the data coordinates in the original uncalibrated axis
```

PeterC-DLS commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+    </group>
+    <group name="coefficients" type="NXparameters">
+        <doc>
+             For non-linear energy calibrations, e.g. in a TOF, a polynomial function is fit
```
```suggestion
             For non-linear energy calibrations, e.g. in a TOF, a polynomial function is fitted
```

PeterC-DLS replied:
Explain TOF

PeterC-DLS commented on base_classes/NXinstrument.nxdl.xml:
```diff
@@ -42,6 +42,7 @@
 	        <doc>short name for instrument, perhaps the acronym</doc>
 	    </attribute>
 	</field>
+    <group type="NXactuator"/>
```
Not related to this PR and other ones below too

PeterC-DLS commented:
It's unclear where `NXcalibration` and `NXresolution` fit in to this PR as both are not used anywhere within

lukaspie commented:
> It's unclear where `NXcalibration` and `NXresolution` fit in to this PR as both are not used anywhere within

We had `NXresolution` in `NXinstrument` originally. NIAC didn't like it, so we removed it and it is not used anywhere else right now. Will be used heavily in #1424 however. `NXcalibration` is cited in `NXresolution`.

lukaspie commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+             
+             Use a0, a1, ..., an for the coefficients, corresponding to the values in the coefficients group.
+             
+             Use x0, x1, ..., xn for the nth position in the `original_axis` field.
```
done!

PeterC-DLS commented on base_classes/NXcalibration.nxdl.xml:
```diff
@@ -0,0 +1,207 @@
+    </field>
+    <field name="calibrated_axis" type="NX_FLOAT" units="NX_ANY">
+        <doc>
+             A vector representing the axis after calibration, matching the data length
```
```suggestion
             An array representing the axis after calibration, matching the data length
```

PeterC-DLS replied:
Check elsewhere too

lukaspie replied:
As far as I can tell, these are the only occurences of misuse of vector in this PR

lukaspie commented:
Rebased on top of main in case you are wondering why there are so many "new" commits

PeterC-DLS commented on base_classes/NXhistory.nxdl.xml:
```diff
@@ -0,0 +1,57 @@
+         Ideally, a full report of the previous operations (or links to a chain of operations).
+         Alternatively, notes allow for additional descriptors in any format.
+    </doc>
+    <group type="NXactivity">
```
Should be this have `minOccurs=1` as you want to record a list of activities in this group

lukaspie replied:
In principle yes, but what happens if we start subclassing `NXactivity`, e.g. to `NXmy_special_activity`? I would probably have to add that to `NXhistory` as well, right? Then neither of these must really occur by itself, rather we just want at least one class _extending_ `NXactivity` to occur.

This is why I was advocating in the base class inheritance discussion that if I make use of `NXactivity` anywhere, this should cover `NXactivity` and also all sub-class of `NXactivity`. 

Maybe this discussion does go to far for this PR though, so we can just go with `minOccurs=1` as suggested.

PeterC-DLS replied:
I would have thought that any instance of a subclass of `NXactivity` is considered also as an instance of its base class so fine for one of these activity occurrences.

lukaspie replied:
I agree that it should be like this. Maybe this should be stated more clearly in the documentation because validators of course need to implement this as well.

For now, let's just go with your suggestion here.

lukaspie replied:
I changed it to ´<group type="NXactivity" minOccurs="1">` now.

PeterC-DLS reviewed (approved):
LGTM
